First version of partial evaluator: support for +,-,* operators and associative rules

jeshecdom opened the pull request:
Towards #435.

- First version of partial evaluator. Currently it is able to simplify expressions involving +, -, and *. However, there are some algebraic rules involving those operators that have not been added yet.
- I decided to keep the partial evaluator separate from the constant evaluator because the partial evaluator still does not support certain language constructions, like structs: the partial evaluator needs to create AST nodes out of values, and there is currently no code to construct ASTStruct nodes out of StructValue. This will be added in the future.
- Currently, to call the partial evaluator, one needs to call the function  `partiallyEvalExpression`. In the future, the constant evaluator will call the partial evaluator, because they should give the same result for expressions that can fully evaluate.
- Many algebraic rules are still pending, including the rest of integer operators and boolean operators.
- Still lacks documentation, specially the explanation of how each rule works.
- Since the partial evaluator is currently isolated from the rest of the code, the tests directly check the partial evaluator instead of going through a contract.

- [ ] I have updated CHANGELOG.md
- ~[ ] I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER~
- [X] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [X] I have run all the tests locally and no test failure was reported
- [X] I have run the linter, formatter and spellchecker
- [X] I did not do unrelated and/or undiscussed refactorings

jeshecdom committed:
First commit for partial evaluator.
There are still some TODOs (so, this is only a partial commit):
- Implementation of association rule 3 and simple algebraic rules (to be added to file algebraic.ts)
described in comments of issue: https://github.com/tact-lang/tact/issues/435
- Processing of structs, function calls, id lookups, ternary conditional operator.
- Extensive documentation, specially in the description of what each rule does.
- Extensive testing of the rewrite rules.

jeshecdom committed:
Finished associative rule 3. Refactored Rule type into an interface. Priorities are no longer a property of rules, but are managed by the optimizer.

jeshecdom committed:
Merge branch 'partial_evaluator'. Resolved conflicts in constEval. Added a function to call the partial evaluator in the interpreter to ease testing.

jeshecdom committed:
Added jest test cases for partial evaluator.

jeshecdom committed:
Merge branch 'main' of https://github.com/tact-lang/tact

jeshecdom committed:
Fixes for the linter, spell checker, and code prettifier.

jeshecdom committed:
Fixed another linter problem.

anton-trunov commented:
@jeshecdom awesome stuff! could you please update your PR and pull in the most recent changes from `main` and also resolve the merge conflicts?

anton-trunov commented on src/constEval.ts:
```diff
@@ -97,58 +116,81 @@ function ensureMethodArity(
     op: AstUnaryOperation,
-    operand: AstExpression,
+    valOperand: Value,
+    operandRef: SrcInfo,
```
```suggestion
    operandLoc: SrcInfo,
```

anton-trunov commented on src/constEval.ts:
```diff
@@ -97,58 +116,81 @@ function ensureMethodArity(
     switch (op) {
         case "+":
-            return ensureInt(valOperand, operand.loc);
+            return ensureInt(valOperand, operandRef);
```
we should probably just make the location argument of `ensureInt` optional (the default value would be `DUMMY_LOCATION`), so here the function calls would look like `ensureInt(valOperand)`

jeshecdom committed:
Merge branch 'main' of https://github.com/tact-lang/tact

jeshecdom committed:
Merge branch 'main' into partial_evaluator

anton-trunov commented on src/optimizer/types.ts:
```diff
@@ -0,0 +1,72 @@
+    SrcInfo,
+} from "../grammar/ast";
+
+export type ValueExpression = AstNumber | AstBoolean | AstNull | AstString;
```
Let's rename `ValueExpression` to something different (`AstValue`?), I'm getting headaches from this one :)

anton-trunov commented on src/optimizer/types.ts:
```diff
@@ -0,0 +1,72 @@
+
+export type ValueExpression = AstNumber | AstBoolean | AstNull | AstString;
+
+export const DUMMY_INTERVAL: Interval = {
```
you should be able to reuse the existing dummy definitions from grammar.ts: https://github.com/tact-lang/tact/blob/d8893ca5744772fddabd2071e31931f2b4b83b9b/src/grammar/grammar.ts#L70-L71

anton-trunov commented on src/optimizer/types.ts:
```diff
@@ -0,0 +1,72 @@
+    "user",
+);
+
+export interface ExpressionTransformer {
```
let's not use interfaces in our codebase: they have some weird properties related to 'interface merging"

anton-trunov commented on src/optimizer/util.ts:
```diff
@@ -0,0 +1,303 @@
+        case "method_call":
+            if (ast2.kind === "method_call") {
+                return (
+                    ast1.method.text === ast2.method.text &&
```
Let's use the identifier comparison function `eqNames` here and elsewhere:
```suggestion
                    eqNames(ast1.method, ast2.method) &&
```

anton-trunov commented on src/optimizer/util.ts:
```diff
@@ -0,0 +1,303 @@
+}
+
+// Test equality of ASTExpressions.
+export function areEqualExpressions(
```
this should be moved to `src/grammar/ast.ts` and renamed to `eqExpressions` (there is `eqNames`) to compare identifiers

anton-trunov commented on src/optimizer/util.ts:
```diff
@@ -0,0 +1,303 @@
+    ast1: AstExpression,
+    ast2: AstExpression,
+): boolean {
+    switch (ast1.kind) {
```
I'm wondering if checks like `ast2.kind === "boolean"` would be redundant if we check `ast1.kind === ast2.kind` before the switch?. That would simplify some code

anton-trunov commented on src/test/e2e-emulated/partial-eval.spec.ts:
```diff
@@ -0,0 +1,209 @@
+    { original: "0 * (X * 3)", simplified: "0 * (X * 3)" },
+    { original: "0 * (3 * X)", simplified: "0 * (3 * X)" },
+
+    { original: "X * 0 * 3", simplified: "X * 0" },
```
wondering why this is not `simplified: 0`, is it because the corresponding rule is not implemented yet?

anton-trunov reviewed (changes requested):
At this point I did some superficial code style review mostly, let's merge things, run the linter with the newly introduced checks and then I'll switch to reviewing the algorithm. (from a quick glance it's impressive already)

jeshecdom committed:
Fixed compilation errors after merge.

jeshecdom committed:
Merge branch 'main' into partial_evaluator

jeshecdom committed:
Fixed problems pointed out by linter and prettier.

jeshecdom committed:
- Changed all `Ref` in parameters for `Loc`. (https://github.com/tact-lang/tact/pull/528/files#r1666596822)
- Added `dummySrcInfo` as default parameter values to functions `__evalUnaryOp` and `__evalBinaryOp`. This way I was able to eliminate them and simply leave `evalUnaryOp` and `evalBinaryOp` as exportable functions. (https://github.com/tact-lang/tact/pull/528/files#r1666598821)
- Renamed ValueExpression for AstValue (https://github.com/tact-lang/tact/pull/528/files#r1666602072)
- Used Dummy interval in grammar.ts (https://github.com/tact-lang/tact/pull/528/files#r1666602072)
- Changed interfaces for abstract classes (https://github.com/tact-lang/tact/pull/528/files#r1666604615)
- Renamed `areEqualExpressions` to `eqExpressions`, and moved it to `ast.ts`. Used `eqNames` and moved `ast1.kind === ast2.kind` check before switch. (https://github.com/tact-lang/tact/pull/528/files#r1666609433)

jeshecdom committed:
fixes due to linter and prettier.

jeshecdom committed:
Merge branch 'main' of https://github.com/tact-lang/tact

jeshecdom committed:
Merge branch 'main' into partial_evaluator

anton-trunov commented:
There are merge conflicts

anton-trunov commented on src/grammar/ast.ts:
```diff
@@ -679,6 +679,128 @@ export function __DANGER_resetNodeId() {
+        case "null":
+            return true;
+        case "boolean":
+            return ast1.value === (ast2 as AstBoolean).value;
```
`ast2 as AstBoolean` is a type assertion, so it will fail at run-time if `ast2` is not an actual boolean constant. Basically, you either need to do pattern-matching on both arguments, or catch exceptions and return `false`. Let's refactor this and add some unit tests for `eqExpressions`. We will need this to resolve part of #314 (equivalence checks between ASTs).

anton-trunov commented on src/optimizer/types.ts:
```diff
@@ -0,0 +1,20 @@
+    AstString,
+} from "../grammar/ast";
+
+export type AstValue = AstNumber | AstBoolean | AstNull | AstString;
```
Would it make sense to move this to `ast.ts`?

anton-trunov commented on src/optimizer/util.ts:
```diff
@@ -0,0 +1,167 @@
+        return isValue(ast.left) && !isValue(ast.right);
+    } else {
+        return false;
+    }
```
```suggestion
    return ast.kind === "op_binary" && isValue(ast.left) && !isValue(ast.right)
```

anton-trunov commented on src/constEval.ts:
```diff
@@ -645,25 +702,28 @@ function interpretEscapeSequences(stringLiteral: string) {
     );
 }
 
+function lookupID(ast: AstId, ctx: CompilerContext): Value {
```
```suggestion
function lookupName(ast: AstId, ctx: CompilerContext): Value {
```

this is to make it in one style with `eqNames`

btw, thanks for factoring out this as a standalone function: this looks much better than being inlined

anton-trunov commented on src/constEval.ts:
```diff
@@ -681,9 +741,9 @@ export function evalConstantExpression(
             return ensureString(interpretEscapeSequences(ast.value), ast.loc);
         case "op_unary":
-            return evalUnaryOp(ast.op, ast.operand, ast.loc, ctx);
+            return fullyEvalUnaryOp(ast.op, ast.operand, ast.loc, ctx);
```
maybe we should rename everything to `fullyEvalX` in the `evalConstantExpression` function? for consistency reasons, like `evalMethod` -> `fullyEvalMethod`, etc. We can put a comment that it's ongoing work to make all of those actually fully evaluate

jeshecdom committed:
- Added try/catch in switch of eqExpressions in ast.ts.
- Added test cases for eqExpressions function in ast.ts.
- Changed if statement for conditional expressions in optimizer/util.ts.
- Changed lookupID for lookupName and evalX for fullyEvalX in constEval.ts.

jeshecdom committed:
Merge branch 'main' of https://github.com/tact-lang/tact

jeshecdom committed:
Merge branch 'main' into partial_evaluator

jeshecdom committed:
Forgot to add test cases for initOf.

jeshecdom committed:
Fixed problems indicated by knit.

anton-trunov commented on src/grammar/ast.ts:
```diff
@@ -679,4 +679,316 @@ export function __DANGER_resetNodeId() {
+    } catch (e) {
+        // In principle, the assertions "as Ast___" should not fail
+        // because ast1 and ast2 have the same kind inside the switch.
+        return false;
```
ah, sorry, I missed you had

```
if (ast1.kind !== ast2.kind) {
        return false;
    }
```

above

`try-catch` isn't needed here, my bad

anton-trunov commented on src/grammar/ast.ts:
```diff
@@ -679,4 +679,316 @@ export function __DANGER_resetNodeId() {
+    return true;
+}
+
+export function traverse(node: AstNode, callback: (node: AstNode) => void) {
```
`traverse` has been moved to `grammar/iterators.ts` recently

anton-trunov commented on src/optimizer/util.ts:
```diff
@@ -0,0 +1,163 @@
+import { Value } from "../types/types";
+import { AstValue } from "./types";
+
+export function isValue(ast: AstExpression): boolean {
```
let's move `isValue` to `ast.ts`

anton-trunov commented on src/test/e2e-emulated/expr-equality.spec.ts:
this test should be moved to `src/grammar/test`, because it does not use the Sandbox library to run compiled contracts in a simulated environment (that's what `e2e-emulated` means)

jeshecdom committed:
- Moved `isValue` to ast.ts.
- Moved AstValue to ast.ts.
- Removed try/catch from eqExpressions in ast.ts.
- Removed `traverse` from ast.ts since it has been already moved to iterators.ts
- Moved expr-equality.spec.ts and partial-eval.spec.ts to src/grammar/test
- Added test cases for `isValue' function in ast.ts in src/grammar/test/expr-is-value.spec.ts.

jeshecdom committed:
Merge branch 'main' of https://github.com/tact-lang/tact

jeshecdom committed:
Merge branch 'main' into partial_evaluator

jeshecdom committed:
Fixed compilation error after merge with main.
